### PR TITLE
Remove default `sf::Sprite` constructor

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

Continuation of the new precedent established in #2486.

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
What other types may need to get this treatment? The pattern is that types that hold a raw pointer to a resource ought to not have a default constructor.

`sf::Sound` is a potential candidate for this change. The problem with `sf::Sound` is that when copied that pointer to the sound buffer goes null so to do this we'd have to change the copy semantics of `sf::Sound`. Maybe we remove copy semantics in favor of making it a move-only type to address this.

**Comment by Bromeon** [comment]
> What other types may need to get this treatment? The pattern is that types that hold a raw pointer to a resource ought to not have a default constructor.

Possibly some of shapes?

_[CircleShape](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1CircleShape.php#aaebe705e7180cd55588eb19488af3af1) (float radius=0, std::size_t pointCount=30)
[RectangleShape](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1RectangleShape.php#a83a2be157ebee85c95ed491c3e78dd7c) (const [Vector2f](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1Vector2.php) &size=[Vector2f](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1Vector2.php)(0, 0))_

Are these meaningful defaults? Not necessarily.

But it's important that we're all fully aware that we _do_ restrict quite some use cases by not having things default-constructible, like:

* `std::array` / raw arrays
* `std::vector::resize`
* `std::map::operator[]`
* conditional initialization with `if/else`
* ...

---


> `sf::Sound` is a potential candidate for this change. The problem with `sf::Sound` is that when copied that pointer to the sound buffer goes null so to do this we'd have to change the copy semantics of `sf::Sound`.

It doesn't seem to reset to null?

https://github.com/SFML/SFML/blob/21af6fecf142882b08e1f2aa006bb406a3631e96/src/SFML/Audio/Sound.cpp#L50-L55
https://github.com/SFML/SFML/blob/21af6fecf142882b08e1f2aa006bb406a3631e96/src/SFML/Audio/Sound.cpp#L89-L102

This auto-attach/detaqch seems to me more like something where we could unify the behavior of resources. It can be surprising that sound auto-unlikes itself upon destruction, while others just run into UB. I understand it's probably because sounds run continuously on a separate thread and sounds need to be centrally managed, but still...

---

> Maybe we remove copy semantics in favor of making it a move-only type to address this.

That seems too restrictive to me. These "frontend" classes (sprite, sound, text) are relatively lightweight and I don't see why we should forbid copying.

**Comment by ChrisThrasher** [comment]
> It doesn't seem to reset to null?

https://github.com/SFML/SFML/blob/21af6fecf142882b08e1f2aa006bb406a3631e96/src/SFML/Audio/Sound.cpp#L172

**Comment by ChrisThrasher** [comment]
> > What other types may need to get this treatment? The pattern is that types that hold a raw pointer to a resource ought to not have a default constructor.
> Possibly some of shapes?
> 
> [CircleShape](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1CircleShape.php#aaebe705e7180cd55588eb19488af3af1) (float radius=0, std::size_t pointCount=30)
> [RectangleShape](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1RectangleShape.php#a83a2be157ebee85c95ed491c3e78dd7c) (const [Vector2f](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1Vector2.php) &size=[Vector2f](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1Vector2.php)(0, 0))
> 
> Are these meaningful defaults? Not necessarily.

If a type can only reach a given state through its default constructor, that default constructor maybe shouldn't exist. For example, a `std::vector` is default constructed with zero elements, but it's still valid to call `.clear()` to get back to that state later. Seems reasonable to me. 

However, an `sf::Sprite` can only have a null texture upon default construction and not through any calls to `setTexture`. The fact that `setTexture` takes a const reference implies that we don't want users to nullify that texture later. It implies to me that having a null texture was never an intentionally valid state but rather a quirk of development that was never second-guessed. This argument probably applies to `sf::Sound` but apparently details of OpenAL complicate things so more research will be required.

By that logic I think it's fine to let an `sf::CircleShape` or `sf::RectangleShape` default construct with a zero size because that's a state you can already achieve in the API. We wrote an API that says "zero size is valid" so it's also valid to have a default constructor that takes advantage of that possibility. 

TL;DR
* Forcing users to specify some initial elements would be annoying when an empty `std::vector` is already a permissible state to put a vector into.
* Forcing users to specify a texture when constructing an `sf::Sprite` makes sense because there is no way to nullify that texture later.
* Forcing users to specify a radius/size on a `sf::RectangleShape` or `sf::CircleShape` would be annoying when it's already valid to set those values to all zeros.

**Comment by Bromeon** [comment]
> If a type can only reach a given state through its default constructor, that default constructor maybe shouldn't exist. For example, a `std::vector` is default constructed with zero elements, but it's still valid to call `.clear()` to get back to that state later. Seems reasonable to me.
> 
> However, an `sf::Sprite` can only have a null texture upon default construction and not through any calls to `setTexture`. The fact that `setTexture` takes a const reference implies that we don't want users to nullify that texture later.

This is a great observation, makes total sense 👍 

I'm not sure if it alone is a sufficient criterion though. `std::vector` being empty by default is _not only_ reasonable because one can clear it, but also because it's a very useful default. In a ton of scenarios, one adds only elements over time (like bullet entities in a game), and starting out empty is desired.

In contrast, a rectangle or circle with 0 extents is a degenerate case and almost never what someone wants -- even in the few hard-to-construct scenarios where someone intentionally needs this, it's always possible to explicitly provide such arguments.

To be clear, I'm not saying `sf::Shape` derivates cannot have default constructors. But these default constructors help mainly to get around technical limitations (e.g. using `std::array`), but are semantically useless.

File: src/SFML/Graphics/Sprite.cpp [hunk 1]
```diff
@@ -133,14 +129,11 @@ FloatRect Sprite::getGlobalBounds() const
...
-        RenderStates statesCopy(states);
+    RenderStates statesCopy(states);
 
-        statesCopy.transform *= getTransform();
-        statesCopy.texture = m_texture;
-        target.draw(m_vertices, 4, PrimitiveType::TriangleStrip, statesCopy);
-    }
+    statesCopy.transform *= getTransform();
+    statesCopy.texture = m_texture;
+    target.draw(m_vertices, 4, PrimitiveType::TriangleStrip, statesCopy);
```
Comment (vittorioromeo) [inline src/SFML/Graphics/Sprite.cpp:136]: Should we `assert(m_texture)`?

Comment (ChrisThrasher) [reply src/SFML/Graphics/Sprite.cpp:136]: Under what circumstances could that pointer become null? I used to have that check in there then removed it because as far as I can tell it's impossible for it to get to that state.

Comment (vittorioromeo) [reply src/SFML/Graphics/Sprite.cpp:136]: > as far as I can tell it's impossible for it to get to that state

That seems a use case for `assert` :P

Comment (ChrisThrasher) [reply src/SFML/Graphics/Sprite.cpp]: `assert`s are so cheap that I won't protest its addition but I'm still really curious how under any circumstances other than `memcpy` shenanigans that pointer could become null.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Graphics/Sprite.cpp [hunk 2]
```diff
@@ -133,14 +129,13 @@ FloatRect Sprite::getGlobalBounds() const
 ////////////////////////////////////////////////////////////
 void Sprite::draw(RenderTarget& target, const RenderStates& states) const
 {
-    if (m_texture)
-    {
-        RenderStates statesCopy(states);
+    assert(m_texture);
```
Comment (vittorioromeo) [inline src/SFML/Graphics/Sprite.cpp:132]: We need to include `<cassert>` if we want to keep this assert.

**Review by vittorioromeo — CHANGES_REQUESTED** [review]

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by MarioLiebisch** [comment]
> In contrast, a rectangle or circle with 0 extents is a degenerate case and almost never what someone wants -- even in the few hard-to-construct scenarios where someone intentionally needs this, it's always possible to explicitly provide such arguments.

I think it would be reasonable to force an explicit radius in many cases (but not in all of them; as a member variable it would be annoying), but I'd also consider that even with the defaulted radius of 0 the object in itself is still valid. The other calls are not failing after all.

Also, using this approach, one could even question if it should be possible to default a (geometrical) vector to all zeroes, because a zero length vector would be a special edge case just like a zero radius circle.

**Comment by Bromeon** [comment]
> I think it would be reasonable to force an explicit radius in many cases (but not in all of them; as a member variable it would be annoying)

Is "annoying" a good enough reason though? It's also annoying to be forced to initialized `sf::Text` in the constructor; there may not be a `sf::Font` available at the time. On the other hand, it gives you the guarantee that whenever you see a `sf::RectangleShape` object, it is already initialized.

> Also, using this approach, one could even question if it should be possible to default a (geometrical) vector to all zeroes, because a zero length vector would be a special edge case just like a zero radius circle.

Fully disagree, zero vectors are geometrically meaningful for a huge number of cases: coordinate origin, zero offset, zero velocity, etc. They are a _very_ reasonable default value, the same way that 0 is for `int`.

**Comment by MarioLiebisch** [comment]
But just like a zero velocity make sense (even if there's no direction), wouldn't a zero radius circle make sense? A vector is defined by length and orientation, yet a zero vector lacks a clear orientation. A circle is defined by a radius (and number of segments, technically), so would a zero radius circle still be a circle? Maybe getting a bit too philosophical…😉

I think `sf::Sprite` is a completely different case. It's a visual representation of a part of a texture, so without a texture it doesn't make any sense, unless you want to do drawing in shaders only. So maybe even introduce a `sf::Texture::None` static for cases where the texture isn't existing?

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by Hapaxia** [comment]
> > It doesn't seem to reset to null?
> 
> https://github.com/SFML/SFML/blob/21af6fecf142882b08e1f2aa006bb406a3631e96/src/SFML/Audio/Sound.cpp#L172

This is removing the current buffer from the sound to which the new sound has been applied so that the new sound's buffer can replace it (since we're copying from that sound):
https://github.com/SFML/SFML/blob/21af6fecf142882b08e1f2aa006bb406a3631e96/src/SFML/Audio/Sound.cpp#L176-L177

**Review by vittorioromeo — APPROVED** [review]
LGTM

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]
